### PR TITLE
[4.0] Add Controller suffix to extension controllers

### PR DESCRIPTION
**Pull request by laoneo**
### Summary of Changes
This are the controller changes to give the component MVC classes more meaningful names. Controllers will have now `Controller` as suffix to their class name. So `Joomla\Component\Banners\Administrator\Controller\Banners` becomes `Joomla\Component\Banners\Administrator\Controller\BannersController`.

Additionally the default controller which is called just `Controller` gets the name `DisplayController`, to better reflect whet the class does.

### Testing Instructions
- Create a banner in the back end
- Create an article in the back end

### Expected result
All is working as expected.

### Actual result
All is working as expected.

### Documentation Changes Required
If there is already documentation written for the namespaced controllers, then it needs to be updated.

### CODE REVIEW

**Comment by laoneo** [comment]
Accidentally deleted branch. Thanks reflog!!

File: libraries/src/Dispatcher/Dispatcher.php [hunk 1]
```diff
@@ -149,7 +149,7 @@ public function dispatch()
 		else
 		{
 			// Do we have a controller?
-			$controller = $this->input->get('controller', 'controller');
+			$controller = $this->input->get('controller', 'display');
```
Comment (wilsonge) [inline libraries/src/Dispatcher/Dispatcher.php]: Why are we changing this?

File: libraries/src/Controller/Form.php [hunk 2]
```diff
@@ -93,12 +93,13 @@ public function __construct($config = array(), MvcFactoryInterface $factory = nu
 		{
 			$r = null;
 
-			if (!preg_match('/(.*)Controller(.*)/i', get_class($this), $r))
+			if (!preg_match('/(.*)Controller\\\\(.*)/i', get_class($this), $r))
```
Comment (wilsonge) [inline libraries/src/Controller/Form.php]: This is going to break non-namespaced controllers

File: libraries/src/Dispatcher/Dispatcher.php [hunk 3]
```diff
@@ -199,7 +199,19 @@ public function getController($name, $client = null, $config = array())
...
-		$controllerClass = $namespace . $client . '\\Controller\\' . ucfirst($name);
+		$controllerClass = $namespace . $client . '\\Controller\\' . ucfirst($name) . 'Controller';
+
+		// @todo Remove me when core extensions are converted
+		if (!class_exists($controllerClass))
+		{
+			$controllerClass = $namespace . $client . '\\Controller\\' . ucfirst($name);
+		}
+
+		// @todo Remove me when core extensions are converted
```
Comment (wilsonge) [inline libraries/src/Dispatcher/Dispatcher.php:210]: Let's just do this now?

Comment (laoneo) [reply libraries/src/Dispatcher/Dispatcher.php:210]: I wanted to make it reviewable. But I will change them after the merge.

File: libraries/src/Controller/Form.php (hunk 2 above)
Comment (laoneo) [reply libraries/src/Controller/Form.php]: True, will fix it.

Comment (laoneo) [reply libraries/src/Controller/Form.php]: Fixed it, tested it with web links and it worked

File: libraries/src/Dispatcher/Dispatcher.php (hunk 3 above)
Comment (laoneo) [reply libraries/src/Dispatcher/Dispatcher.php:210]: And first, #17682 needs to get it, otherwise I git tons of merge conflicts

**Pull request merged**